ecl/hql/hqlutil.cpp

```diff
@@ -2645,6 +2645,49 @@ void checkSelectConsistency(IHqlExpression * expr)
 
 //---------------------------------------------------------------------------
 
+static HqlTransformerInfo parameterDependencyCheckerInfo("ParameterDependencyChecker");
+class ParameterDependencyChecker  : public NewHqlTransformer
+{
+public:
+    ParameterDependencyChecker() : NewHqlTransformer(parameterDependencyCheckerInfo), foundParameter(false)
+    {
+    }
+
+    virtual void analyseExpr(IHqlExpression * expr)
+    {
+        if (expr->isFullyBound() || alreadyVisited(expr) || foundParameter)
```

Comment:
alreadyVisited() is there so that when you are traversing a graph you don't walk a node more than once.  It is a fairly standard call/pattern for each of the transformers because without it you can get a combinatorial explosion of the number of nodes visited.
Essentially whether this expression really uses a parameter is not context dependent - so we only check it once.